webapp/src/ts/modals/edit-message-group/edit-message-group.component.ts

```diff
@@ -0,0 +1,154 @@
+import { AfterViewChecked, AfterViewInit, Component, OnDestroy, OnInit } from '@angular/core';
+import * as moment from 'moment';
+import { BsModalRef } from 'ngx-bootstrap/modal';
+import { v4 as uuid } from 'uuid';
+
+import { EditGroupService } from '@mm-services/edit-group.service';
+import { SettingsService } from '@mm-services/settings.service';
+import { MmModalAbstract } from '@mm-modals/mm-modal/mm-modal';
+
+@Component({
+  selector: 'edit-message-group',
+  templateUrl: './edit-message-group.component.html',
+})
+export class EditMessageGroupComponent extends MmModalAbstract
+  implements AfterViewInit, OnInit, AfterViewChecked, OnDestroy {
+
+  constructor(
+    bsModalRef:BsModalRef,
+    private editGroupService:EditGroupService,
+    private settingsService:SettingsService,
+  ) {
+    super(bsModalRef);
+  }
+
+  model:any = { group: { rows: [] } };
+
+  private shouldInitDatePickers = false;
+  private settingsPromise;
+  private datePickers = {};
+
+  ngOnInit() {
+    this.getSettings();
+    // task_id variable is only used within this component, to uniquely identify each task in the group
+    this.model?.group?.rows?.forEach(row => row.task_id = uuid());
+  }
+
+  private getSettings() {
+    this.settingsPromise = this.settingsService.get();
+  }
+
+  private getNextHalfHour() {
+    const time = moment()
+      .second(0)
+      .millisecond(0);
+    if (time.minute() < 30) {
+      time.minute(30);
+    } else {
+      time.minute(0);
+      time.add(1, 'hours');
+    }
+    return time;
+  }
+
+  trackBy(index, task) {
+    return task.task_id;
+  }
+
+  private initDatePickers() {
+    this.settingsPromise.then((settings:any) => {
+      $('#edit-message-group input.datepicker').each((index, element) => {
+        const taskId = $(element).data('task-id');
+        if (this.datePickers[taskId]) {
+          // already has datepicker!
+          return;
+        }
+
+        $(element).daterangepicker(
+          {
+            startDate: new Date(this.model.group.rows[index].due),
+            singleDatePicker: true,
+            timePicker: true,
+            applyClass: 'btn-primary',
+            cancelClass: 'btn-link',
+            parentEl: '#edit-message-group',
+            minDate: this.getNextHalfHour(),
+            locale: {
+              format: settings.reported_date_format,
+            },
+          },
+          (date) => {
+            const taskId = $(element).data('task-id');
+            const task = this.model.group.rows.find(task => task.task_id === taskId);
+            task.due = date.toISOString();
+          }
+        );
+        this.datePickers[taskId] = $(element).data('daterangepicker');
+      });
+    });
+  }
+
+  private removeDatePickers(taskId?) {
+    if (taskId) {
+      this.datePickers[taskId]?.remove();
+      delete this.datePickers[taskId];
+      return;
+    }
+
+    Object.keys(this.datePickers).forEach((key) => {
```

Comment:
```suggestion
    Object.keys(this.datePickers).forEach(key => {
```

I see in many parts single arguments in lambdas surrounded with `()`, not an issue though.

Reply:
Yea, when I change a `function(thing) {` I just change it to `(thing) => {`. I mostly leave the brackets in case I have to type the argument later `(thing:<type>)`. 
I don't feel strongly about it one way or the other, in the end.